pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,327 @@
+            ``qml.register_resources``.
+
+    .. seealso:: :func:`~pennylane.register_resources` and :class:`~pennylane.list_decomps`
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+
+        qml.decompositions.enable_graph()
+
+        @qml.register_resources({qml.RZ: 2, qml.RX: 1, qml.GlobalPhase: 1})
+        def my_hadamard1(wires):
+            qml.RZ(np.pi / 2, wires=wires)
+            qml.RX(np.pi / 2, wires=wires)
+            qml.RZ(np.pi / 2, wires=wires)
+            qml.GlobalPhase(-np.pi / 2, wires=wires)
+
+        @qml.register_resources({qml.RZ: 1, qml.RY: 1, qml.GlobalPhase: 1})
+        def my_hadamard2(wires):
+            qml.RZ(np.pi, wires=wires)
+            qml.RY(np.pi / 2, wires=wires)
+            qml.GlobalPhase(-np.pi / 2)
+
+        qml.add_decomps(qml.Hadamard, my_hadamard1, my_hadamard2)
+
+    These two new decomposition rules for ``qml.Hadamard`` will be subsequently stored within the
+    scope of this program, and they will be taken into account for all circuit decompositions
+    for the duration of the session. To add alternative decompositions for a particular circuit
+    as opposed to globally, use the ``alt_decomps`` argument of the ``decompose`` transform.
+
+    .. seealso:: :func:`~pennylane.transforms.decompose`
+
+    """
+    if not all(isinstance(d, DecompositionRule) for d in decomps):
+        raise TypeError(
+            "A decomposition rule must be a qfunc with a resource estimate "
+            "registered using qml.register_resources"
+        )
+    _decompositions[op_type].extend(decomps)
+
+
+def list_decomps(op_type: Type[Operator]) -> list[DecompositionRule]:
+    """Lists all stored decomposition rules for an operator class.
+
+    .. note::
+
+        This function is only relevant when the new experimental graph-based decomposition system
+        (introduced in v0.41) is enabled via ``qml.decompositions.enable_graph()``. This new way of
+        doing decompositions is generally more performant and accommodates multiple alternative
+        decomposition rules for an operator.
+
+    Args:
+        op_type: the operator class to retrieve decomposition rules for.
+
+    Returns:
+        list[DecompositionRule]: a list of decomposition rules registered for the given operator.
+
+    """
+    return _decompositions[op_type][:]
+
+
+def has_decomp(op_type: Type[Operator]) -> bool:
+    """Check whether an operator has decomposition rules defined.
+
+    .. note::
+
+        This function is only relevant when the new experimental graph-based decomposition system
+        (introduced in v0.41) is enabled via ``qml.decompositions.enable_graph()``. This new way of
+        doing decompositions is generally more performant and accommodates multiple alternative
+        decomposition rules for an operator.
+
+    Args:
+        op_type: the operator class to check for decomposition rules.
+
```

Comment:
sorry I forgot to write a docstring for this one!

```suggestion

    Returns:
        Bool: ``True``/``False`` depending on whether the operator in question has decompostion rules defined.
```